zebra/zebra_pm.c

```diff
@@ -0,0 +1,417 @@
+/*
+ * Zebra PM Echo - Zebra Path Monitoring proxy functions
+ * Copyright (C) 6WIND 2019
+ *
+ * This file is part of FRR.
+ *
+ * FRR is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2, or (at your option) any
+ * later version.
+ *
+ * FRR is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+
+#include <sys/un.h> /* for sockaddr_un */
+#include <net/if.h>
+
+#include "pm_lib.h"
+#include "buffer.h"
+#include "command.h"
+#include "if.h"
+#include "network.h"
+#include "ptm_lib.h"
+#include "rib.h"
+#include "stream.h"
+#include "version.h"
+#include "vrf.h"
+#include "vty.h"
+#include "lib_errors.h"
+
+#include "zebra/debug.h"
+#include "zebra/interface.h"
+#include "zebra/zebra_errors.h"
+#include "zebra/zebra_pm.h"
+#include "zebra/zserv.h"
+#include "zebra_vrf.h"
+#include "zebra/zebra_memory.h"
+#include "zebra/zebra_router.h"
+/*
+ * Data structures.
+ */
+struct pm_process {
+	struct zserv *pp_zs;
+	pid_t pp_pid;
+
+	TAILQ_ENTRY(pm_process) pp_entry;
+};
+TAILQ_HEAD(pm_ppqueue, pm_process) pm_ppqueue;
+
+DEFINE_MTYPE_STATIC(ZEBRA, ZEBRA_PM_PROCESS,
+		    "PM process registration table.");
+
+/*
+ * Prototypes.
+ */
+static struct pm_process *pp_new(pid_t pid, struct zserv *zs);
+static struct pm_process *pp_lookup_byzs(struct zserv *zs);
+static void pp_free(struct pm_process *pp);
+static void pp_free_all(void);
+
+static void zebra_pm_send_pmd(struct stream *msg);
+static void zebra_pm_send_clients(struct stream *msg);
+static int _zebra_pm_client_deregister(struct zserv *zs);
+static void _zebra_pm_reroute(struct zserv *zs, struct zebra_vrf *zvrf,
+			       struct stream *msg, uint32_t command);
+
+
+/*
+ * Process PID registration.
+ */
+static struct pm_process *pp_new(pid_t pid, struct zserv *zs)
```

Comment:
I think we could move this code somewhere else and use it for `bfdd` AND `pm`. This way you can get your line count down and make a lot of maintainers happy :) .